src/java/org/apache/cassandra/schema/IndexMetadata.java

```diff
@@ -106,20 +113,53 @@ public static IndexMetadata fromIndexTargets(List<IndexTarget> targets,
         return new IndexMetadata(name, newOptions, kind);
     }
 
-    public static String generateDefaultIndexName(String table, ColumnIdentifier column)
+    /**
+     * Generates a default index name from the table and column names.
+     * Characters other than alphanumeric and underscore are removed.
+     * Long index names are truncated to fit the length allowing constructing filenames.
+     *
+     * @param table  the table name
+     * @param column the column identifier. Can be null if the index is for the primary key.
+     * @return the generated index name
+     */
+    public static String generateDefaultIndexName(String table, @Nullable ColumnIdentifier column)
     {
-        return PATTERN_NON_WORD_CHAR.matcher(table + '_' + column.toString() + "_idx").replaceAll("");
+        String indexNameUncleaned = table;
+        if (column != null)
+            indexNameUncleaned += '_' + column.toString();
+        String indexNameUntrimmed = PATTERN_NON_WORD_CHAR.matcher(indexNameUncleaned).replaceAll("");
+        String indexNameTrimmed = indexNameUntrimmed
+                                  .substring(0,
+                                             Math.min(calculateGeneratedIndexNameMaxLength(),
+                                                      indexNameUntrimmed.length()));
+        return indexNameTrimmed + INDEX_POSTFIX;
     }
 
-    public static String generateDefaultIndexName(String table)
+    /**
+     * Calculates the maximum length of the generated index name to fit file names.
+     * It includes the generated suffixes in account.
+     *
+     * @return the allowed length of the generated index name
+     */
+    private static int calculateGeneratedIndexNameMaxLength()
     {
-        return PATTERN_NON_WORD_CHAR.matcher(table + '_' + "idx").replaceAll("");
+        // Speculative assumption that uniqueness breaker will fit into 999.
+        // The value is used for trimming the index name if needed.
+        // Introducing validation of index name length is TODO for CNDB-13198.
+        int uniquenessSuffixLength = 4;
+        int indexNameAddition = uniquenessSuffixLength + INDEX_POSTFIX.length();
+        int allowedIndexNameLength = Version.calculateIndexNameAllowedLength();
```

Comment:
Isn't this calling SAI-specific rules about index names for all index implementations?

Reply:
Can you expand what you are meaning by the question?
Version is located in SAI.

Reply:
Legacy index also uses index name for file creation, but adds only one char.

Reply:
The logic is in SAI, but the explicit call to `org.apache.cassandra.index.sai.disk.format.Version#calculateIndexNameAllowedLength` makes this SAI-specific, and it applies SAI validation rules to all index implementations. 

I think we should apply the pattern used for validating the index options, where every index implementation can define a `Index.validateOptions` static method, to be called with reflection by `IndexMetadata#validateCustomIndexOptions`. That way, the validation of options is effectively delegated to the index without the need to create the index beforehand. We could apply the same pattern here, so each index implementation can provide its own max index name length.

Reply:
Do you want this solution to be done within this PR or after this PR? At least for me at this stage it's not trivial to find all files creations to get correct calculation of file names for different indexes.
In general I fully agree that having an interface for each index implementation to implement the calculation or make an index name truncation or index name generation is cleaner. However, I don't think I will be ready with the clean enough solution for the release deadline.

Reply:
It’s fine with me to leave it as it is. I would just left a TODO comment pointing to a follow-up ticket.

Reply:
I created https://github.com/riptano/cndb/issues/13240
Adding a comment with TODO will add new issue in Sonar. IMHO it's not critical to have this todo comment.

Reply:
Great, thanks. I think Sonar is there to help us, and if we think we need a TODO comment because we have left something to do separately, Sonar complaints about unfinished stuff shouldn't stop us, IMO. Or you can always add a non-TODO comment simply referencing CNDB-13240, and Sonar will never know.

Reply:
I added this work and reference in the function comment.